### PR TITLE
executor: close indexlookup executor in background

### LINKED ISSUE #29716: executor: SQL after IndexLookUpExecutor may be slow
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

- 1. Create a table with a non-unique index, generate about 10,000,000 rows(I use this [script](https://gist.github.com/you06/c4f5032b6492864e72afea2119207c85)).
- 2. Make an indexlookup query with limit, once it's done, do `select 1;`.
- 3. Repeat step2.

### 2. What did you expect to see? (Required)

The `select 1;` query should be fast, at least less than 10ms.

### 3. What did you see instead (Required)

```sql
MySQL [test]> select * from t force index(i) where c2 > 1 and c2 < 12000000 and c3 in (3) order by c2 desc limit 11; select 1;
+----------+----------+------+
| c1       | c2       | c3   |
+----------+----------+------+
|  2925230 | 11999949 |    3 |
|  4950694 | 11999947 |    3 |
| 11778881 | 11999851 |    3 |
|  2244706 | 11999851 |    3 |
| 10488089 | 11999821 |    3 |
|  5891917 | 11999812 |    3 |
| 11706844 | 11999801 |    3 |
| 10178503 | 11999619 |    3 |
| 13321140 | 11999576 |    3 |
|  5117855 | 11999563 |    3 |
|  9638867 | 11999561 |    3 |
+----------+----------+------+
11 rows in set (0.00 sec)

+---+
| 1 |
+---+
| 1 |
+---+
1 row in set (0.49 sec)
```

The `select 1;` costs about half a second.

### 4. What is your TiDB version? (Required)

```
Release Version: v5.4.0-alpha-87-g58c02d5bb
Edition: Community
Git Commit Hash: 58c02d5bb7b25aa810c8279c5eb7de95a0362356
Git Branch: master
UTC Build Time: 2021-11-11 11:47:06
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by you06**
Signed-off-by: you06 <you1474600@gmail.com>

### What problem does this PR solve?

Issue Number: close #29716

Problem Summary:

When closing the IndexLookUpExecutor, it'll wait until all on-flying requests are done, which will increase the latency of the next SQL execution, this PR makes the close jobs executed in the background.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: distsql/select_result.go [hunk 1]
```diff
@@ -196,6 +199,11 @@ func (r *selectResult) fetchResp(ctx context.Context) error {
 		if err != nil {
 			return errors.Trace(err)
 		}
+		r.mu.Lock()
```
Comment (fuzhe1989) [inline distsql/select_result.go:202]: guess it's better to extract the for-body as a function and defer `r.mu.Unlock()` (if still use `sync.Mutex`).

File: distsql/select_result.go [hunk 2]
```diff
@@ -153,6 +154,8 @@ type selectResult struct {
 	memTracker       *memory.Tracker
 
 	stats *selectResultRuntimeStats
+	mu    sync.Mutex
+	exit  bool
```
Comment (fuzhe1989) [inline distsql/select_result.go:158]: could we use sync/atomic here?

Comment (you06) [reply distsql/select_result.go:158]: The atomic seems not enough here. When there is a response, if we check the atomic variable and make sure it's closed yet, then we start updating the consuming memory and copr stats, just before the update is executing, the result is closed and there may be a nil dereference.

File: distsql/select_result.go (hunk 1 above)
Comment (you06) [reply distsql/select_result.go]: Done.

**Comment by you06** [comment]
@fuzhe1989 @wjhuang2016 can you help review this PR?

**Comment by fuzhe1989** [comment]
> @fuzhe1989 @wjhuang2016 can you help review this PR?

Guess you need to find another reviewer in that my LGTM has no effect ...

**Comment by you06** [comment]
@crazycs520 @ichn-hu @tiancaiamao PTAL

File: executor/distsql.go [hunk 3]
```diff
@@ -582,6 +594,19 @@ func (e *IndexLookUpExecutor) startIndexWorker(ctx context.Context, workCh chan<
...
+				finished = true
+			default:
+			}
+			if finished {
+				result.Close()
+				e.mu.Unlock()
+				break
+			}
+			e.pendingIndexResults = result
+			e.mu.Unlock()
```
Comment (tiancaiamao) [inline executor/distsql.go:609]: What's the difference check again here between checking in the next loop at line 578?

File: executor/distsql.go [hunk 4]
```diff
@@ -676,17 +705,33 @@ func (e *IndexLookUpExecutor) Close() error {
...
-	e.tblWorkerWg.Wait()
 	e.finished = nil
 	e.workerStarted = false
+	if e.pendingIndexResults != nil {
+		e.pendingIndexResults.Close()
+	}
+	for _, worker := range e.tableWorkers {
+		worker.mu.Lock()
+		if worker.tableReader != nil {
+			worker.tableReader.Close()
```
Comment (tiancaiamao) [inline executor/distsql.go:718]: Close table reader before index worker, the index task may block on sending to the table task worker ... and then goroutine leak.

File: executor/distsql.go [hunk 5]
```diff
@@ -676,17 +705,33 @@ func (e *IndexLookUpExecutor) Close() error {
...
+		}
+		worker.mu.Unlock()
+	}
 	e.memTracker = nil
 	e.resultCurr = nil
+	resultCh, idxWorkerWg, tblWorkerWg := e.resultCh, e.idxWorkerWg, e.tblWorkerWg
+	e.mu.Unlock()
+	// Drain the resultCh and discard the result, in case that Next() doesn't fully
+	// consume the data, background worker still writing to resultCh and block forever.
+	go func() {
```
Comment (tiancaiamao) [inline executor/distsql.go:729]: If the executor do not wait for task to exit, I'm afraid there will be some data race or chunk reuse problems....

File: executor/distsql.go [hunk 6]
```diff
@@ -1179,11 +1226,17 @@ func (w *tableWorker) compareData(ctx context.Context, task *lookupTableTask, ta
...
 	tableReader, err := w.idxLookup.buildTableReader(ctx, task)
+	w.mu.Lock()
+	w.tableReader = tableReader
+	w.mu.Unlock()
 	if err != nil {
 		logutil.Logger(ctx).Error("build table reader failed", zap.Error(err))
 		return err
 	}
-	defer terror.Call(tableReader.Close)
+	defer terror.Call(func() error {
```
Comment (tiancaiamao) [inline executor/distsql.go:1236]: What's the difference before & after this change?

**Review by tiancaiamao — COMMENTED** [review]
The code is fragile... in history we see so many bugs related to goroutine block & resource leak & query hang ... I'm not quite confident with the changes.

**Comment by you06** [comment]
Close this PR since the change is too risky.

**Comment by you06** [comment]
The problem is fixed by #29894.

**Pull request closed without merging**